Nullable.md

```diff
@@ -0,0 +1,442 @@
+indirection has many implications because it ensures that
+`T <: Union{Some{T}, Void}` is not true, breaking many of the patterns we noted
+with `Union{T, Void}`.
+
+For example, because a wrapper is employed, no functions with methods defined
+on `T` will work on objects of type `Union{Some{T}, Void}` unless those methods
+are defined for `f(x::Some{T})` and `f(x::Void)`. For example, the existence of
+`f(x::Int)` does not imply that `f(x::Some{Int})` or `f(x::Void)` will be
+defined.
+
+This limitation can also be framed as a benefit because it ensures that
+multiple methods can co-exist without the specificity concerns raised for
+`Union{T, Void}`.
+
+For example, it would be possible to write code that hits all branches of the
+following method definitions:
+
+```
+f(x::Int) = 1
+f(x::Union{Some{Int}, Void}) = isa(x, Void) ? 2 : 3
+```
+
+To repeat this point in other terms,
+the use of `Some{T}` instead of `T` provides a way of having the `T` and `?T`
+types co-exist without specificity issues, but it simultaneously ensures
+that no existing methods for `T` automatically apply to `?T`. It provides
+arguably cleaner semantics (by making `T` and `?T` disjoint) at the cost of
+making the non-null case for nullable types less convenient to work with.
+
+As such, this implementation of `?T` would require that we push users towards
+the consistent usage of specialized lifting machinery that knows how to handle
+`f(x::Union{Some{T}, Void})` given only knowledge of how `f(x::T)` works. For
+one approach to this problem, see the later section on lifting functions from
+non-nullable types to nullable types.
+
+Because users will have to work with lifted functions, this approach makes it
+more difficult for users to write code that works for the non-null case, but
+does not work for the null case. But the possibility of defining methods for
+only one of `Some{T}` or `Void` means that a totality checker would still be
+needed because users can always define only one of `f(x::Some{T})` and
+`f(x::Void)`. (We should note that the current implementation of `Nullable{T}`
+also does not ensure totality.)
+
+Another point in favor of `Some{T}` is that it is sometimes useful to
+distinguish between two confusable outcomes in cases where a method returns
+something like `Union{Some{Void}, Void}` -- as might happen in a hash table
+implementation that returns `Some(nothing)` when a key is present and
+associated with `nothing` as a value, but returns `nothing` when a key is not
+present. This is another implication of working with disjoint types for `T` and
+`?T`.
+
+On the other hand, the use of `Some{T}` raises issues about how we expect `T`,
+`Some{T}` and `Void` to interact. For example, consider the following code:
+
+```
+function sum(xs::Vector{Union{Some{Int}, Void}})
+    s = 0
+    for x in xs
+        s += x
+    end
+    return x
+end
+```
+
+What type does this return when `xs = ?Int[]`? What type does it return when
```

Comment:
Could you illustrate the alternative solutions? It's not completely clear to me what problem you're describing.